### PR TITLE
Add moderation actions table

### LINKED ISSUE #297: Create mod actions database table
There are quite some features being implemented right now which are blocked by having a way to retrieve moderative actions that have been executed in the past:

* Warn command #63 (to have a place to store the reason)
* Audit command #249 (to get a list of all actions for a user)
* Temporary bans #267 (to undo the action)
* Mute command #252 (to reapply the role after leaving)
* Actual author in log messages instead of bot #280 (to figure out the actual author behind the command)

### PR DESCRIPTION
**Pull request by Zabuzard**
# Overview

Implements and closes #297. This is a **crucial pre-requisite that blocks** a whole list of features that are currently either being developed or in planning:

* Warn command #249 (to have a place to store the reason)
* Audit command #301 (to get a list of all actions for a user)
* Temporary bans #267 (to undo the action)
* Mute command #252 (to reapply the role after leaving)
* Actual author in log messages instead of bot #280 (to figure out the actual author behind the command)

This feature basically logs each moderation action (ban, kick, mute, ...) in a database table, with all necessary information such as target, author, timestamp, reason and more. This can then be used by other classes later to retrieve information (see list of blocked features).

# Details

The implementation adds a new database table `moderation_actions`:
```sql
CREATE TABLE moderation_actions
(
    case_id           INTEGER   NOT NULL PRIMARY KEY AUTOINCREMENT,
    issued_at         TIMESTAMP NOT NULL,
    guild_id          BIGINT    NOT NULL,
    author_id         BIGINT    NOT NULL,
    target_id         BIGINT    NOT NULL,
    action_type       TEXT      NOT NULL,
    action_expires_at TIMESTAMP,
    reason            TEXT      NOT NULL
)
```
which is hidden behind a helper class `ModerationActionsStore` with retrieval methods such as:
* `getActionsByTypeAscending`
* `getActionsByTargetAscending`
* `getActionsByAuthorAscending`
* `findActionByCaseId`

(while those methods are not used right now, they will be used very soon by the mentioned features)

to add a new entry, one calls
* `addAction`

Additionally, the store has been integrated into all currently existing commands:
* `BanCommand`
* `UnbanCommand`
* `KickCommand`

so those commands will now also correctly add the entry to the database, when executed.

# Notes

The point in time commands call the `addAction` method as of now is slightly off the actual point in time the action is executed.

Simplified speaking, we have
```java
addAction(...);
guild.ban(...);
```
while the `ban` command is obviously a rest action that might be executed slightly deferred. That said, we could potentially use
```java
guild.ban(...)
  .map(result -> {
    addAction(...);
    return result;
  });
```
in most cases instead (but maybe not all). However, the time will then still be slightly offset (even if its just some milliseconds).

(I wrote it like that to stay consistent with the point in time we have our actual log message being written. So if we should change this, the log also has to go into the `map` call)

**tl;dr**: It is not really possible to get the exact timestamp that the action is executed at. Hence, future methods that are added to `ModerationActionsStore`, such as timestamp-based retrieval (thinking about #280), must be implemented with a time variance in mind (action closest to timestamp ...).

# Checklist

* [x] Create the store
* [x] Integrate the store
* [x] Add a note to the class that timestamps might be slightly off
* [x] Javadoc
* [x] Fix Linter issues
* [x] Testing

### CODE REVIEW

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    public @NotNull List<ActionRecord> getActionsByAuthorAscending(long guildId, long authorId) {
+        return database.read(context -> context.selectFrom(ModerationActions.MODERATION_ACTIONS)
+            .where(ModerationActions.MODERATION_ACTIONS.GUILD_ID.eq(guildId)
+                .and(ModerationActions.MODERATION_ACTIONS.AUTHOR_ID.eq(authorId)))
+            .orderBy(ModerationActions.MODERATION_ACTIONS.ISSUED_AT.asc())
+            .stream()
+            .map(ActionRecord::of)
+            .toList());
+    }
```
Comment (Zabuzard) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java:50]: Think about adding a helper method to get rid of duplication here.

Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java]: Fixed.

**Review by RealYusufIsmail — APPROVED** [review]
I have looked thorough it and all looks good.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java [hunk 2]
```diff
@@ -43,9 +46,10 @@ public UnbanCommand() {
 
         hasRequiredRole = Pattern.compile(Config.getInstance().getHeavyModerationRolePattern())
             .asMatchPredicate();
+        this.actionsStore = actionsStore;
```
Comment (Tais993) [inline application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java:49]:
```suggestion
        this.actionsStore = Objects.requireNonNull(actionsStore);
```

**Review by Tais993 — DISMISSED** [review]

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java [hunk 3]
```diff
@@ -72,9 +76,9 @@ public BanCommand() {
     }
 
     @SuppressWarnings("MethodWithTooManyParameters")
-    private static RestAction<InteractionHook> banUserFlow(@NotNull User target,
-            @NotNull Member author, @NotNull String reason, int deleteHistoryDays,
-            @NotNull Guild guild, @NotNull SlashCommandEvent event) {
+    private RestAction<InteractionHook> banUserFlow(@NotNull User target, @NotNull Member author,
+            @NotNull String reason, int deleteHistoryDays, @NotNull Guild guild,
+            @NotNull SlashCommandEvent event) {
```
[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:81]: It is checked by the system though, no? Its new to me that you should actually trigger a NPE explicitly if you already use `@NotNull` annotations.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java [hunk 4]
```diff
@@ -36,18 +36,22 @@
...
-    public KickCommand() {
+    public KickCommand(@NotNull ModerationActionsStore actionsStore) {
         super(COMMAND_NAME, "Kicks the given user from the server", SlashCommandVisibility.GUILD);
 
         getData().addOption(OptionType.USER, TARGET_OPTION, "The user who you want to kick", true)
             .addOption(OptionType.STRING, REASON_OPTION, "Why the user should be kicked", true);
 
         hasRequiredRole = Pattern.compile(Config.getInstance().getSoftModerationRolePattern())
             .asMatchPredicate();
+        this.actionsStore = actionsStore;
```
[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java:54]: (see other comment)

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+public final class ModerationActionsStore {
+    private final Database database;
+
+    /**
+     * Creates a new instance which writes and retrieves actions from a given database.
+     * 
+     * @param database the database to write and retrieve actions from
+     */
+    public ModerationActionsStore(@NotNull Database database) {
+        this.database = database;
```
[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java:39]: (see other comment)

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java (hunk 2 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java:49]: (see other comment)

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:81]: >>> Its new to me that you should actually trigger a NPE explicitly if you already use @NotNull annotations

Al lot of JDA methods have @NotNull but we still check it, so sounds inconsistent to only check it on their side.

Besides that, you can use a plugin which generates requireNonNull for you, this works in IntelliJ but not sure whenever it works outside of that.

[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:81]: I just havent seen this yet, people explicitly using `requireNonNull` (or other triggers for NPE) on stuff that is marked `@NotNull` already. @marko-radosavljevic , @I-Al-Istannen , do you know more here about convention?

[the remark this answers is not in the document]
Comment (Tais993) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:81]: It has to do with your fail fast principle.

Fail fast, don't wait till you get a random NPE.

If you accept null, it should be `@Nullable` NOT `@NotNull`

[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:81]: > I don't think I understand what you mean, or well it seems like there's some kind of a miscommunication.

I do understand you. I just was under the impression that it also has/does the checks without plugin. Or that its so widely supported across IDEs by default already that people generally dont check it extra again by convention.

I am just asking about the convention right now.

[the remark this answers is not in the document]
Comment (marko-radosavljevic) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:81]: > Like, simply passing a null object to a method annoted with @NotNull already triggers a NPE even if you do not check it yourself in the code, the check is "built-in".

Have you tested this? It's something intellIj does during compilation, if you enable it. Which has several issues, we are building it and running it from gradle, also what about other IDEs, etc.

> When you compile your project, the IDE adds assertions to all methods and parameters annotated with the @NotNull annotation. The assertions will fail if null is passed in code where @NotNull is expected. You can disable this option and configure the list of annotations in the Settings/Preferences dialog Ctrl+Alt+S. Go to Build, Execution, Deployment | Compiler. 

That said, our static checkers support this annotation, so sonarlint fails the build with: [[java:S2637] "@NonNull" values should not be set to null](https://sonarcloud.io/organizations/togetherjava/rules?open=java%3AS2637&rule_key=java%3AS2637).

Also, not sure how other IDEs support this annotation, for example does this work:
image

All in all, not sure. We have a static checking, but I would like to have a runtime check as well. If there is a plugin that does this for gradle, it would be fine with me.

[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:81]: Sounds reasonable. I guess I will just add the checks and keep it in mind for next time. Good spot, thanks.

[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java:81]: I will trust the call-site for private methods though. Otherwise we have to add like 100 checks to every method in the code base, not just the ones you flagged. So for everything that is `private`, I wont start checking everything.

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/KickCommand.java]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+                        .setReason(reason);
+            actionRecord.insert();
+            return actionRecord.getCaseId();
+        });
+    }
+
+    private @NotNull List<ActionRecord> getActionsFromGuildAscending(long guildId,
+            @NotNull Condition condition) {
+        return database.read(context -> context.selectFrom(ModerationActions.MODERATION_ACTIONS)
+            .where(ModerationActions.MODERATION_ACTIONS.GUILD_ID.eq(guildId).and(condition))
```
[the remark this answers is not in the document]
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/ModerationActionsStore.java:147]: done

File: application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java (hunk 2 above)
Comment (Zabuzard) [reply application/src/main/java/org/togetherjava/tjbot/commands/moderation/UnbanCommand.java]: done

**Review by Tais993 — APPROVED** [review]

**Review by RealYusufIsmail — APPROVED** [review]
